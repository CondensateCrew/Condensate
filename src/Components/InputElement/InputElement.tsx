import React, { useState, useEffect } from 'react';
import { UserSignupPosting } from '../../interfaces';

type checkedInputType = 'first-name' | 'last-name' | 'email' | 'password' | 'repeat-password';

interface Props {
  typeInput: checkedInputType,
  user: UserSignupPosting,
  setPassword: (repeatPassword: string) => void
  setUser: (user: UserSignupPosting) => void
  setError: (error: string) => void
}

enum checkType {
  'first-name' = 'firstName',
  'last-name' = 'lastName'
}

enum placeholderText {
  'first-name' = 'Your First Name',
  'last-name' = 'Your First Name',
  'email' = 'email@mail.com',
  'password' = 'Your Password',
  'repeat-password' = 'Repeat Password',
}

const InputElement: React.FC<Props> = ({ typeInput, user, setPassword, setUser, setError }) => {
  const [ textValue, setTextValue ] = useState<string>('');

  const changeUserInfo = (): void => {
    setError('');
    if (typeInput === 'repeat-password') {
      setPassword(textValue);
    }
    else if (typeInput === 'first-name' || typeInput === 'last-name') {
      const userKey = checkType[typeInput];
      setUser({...user, [userKey]: textValue});
    }
    else {
      setUser({...user, [typeInput]: textValue});
    }
  }

  useEffect (changeUserInfo, [ textValue ]);

  const handleChanges = (event: React.ChangeEvent<HTMLInputElement>): void => {
    setTextValue(event.target.value);
  }

  const tp: string = (typeInput === 'password' || typeInput === 'repeat-password')
    ? 'password'
    : 'text';

  const text: string = typeInput.split('-').join(' ');

  return (
    <label htmlFor={typeInput}>
      {text}*
      <input
        id={typeInput}
        type={tp}
        placeholder={placeholderText[typeInput]}
        value={textValue}
        onChange={handleChanges} />
    </label>
  );
}

export default InputElement;
