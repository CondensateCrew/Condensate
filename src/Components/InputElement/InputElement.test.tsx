/**
 * @jest-environment jsdom
 */

import React from 'react';
import InputElement from './InputElement';
import { mount } from 'enzyme';

type checkedInputType = 'first-name' | 'last-name' | 'email' | 'password' | 'repeat-password';

describe('InputElement component', () => {
  let wrapper: any;

  const mockTypeInput: checkedInputType = 'first-name';
  const mockUser = {
    email: '',
    password: '',
    firstName: '',
    lastName: '',
  };
  const mockSetUser = jest.fn();
  const mockSetPassword = jest.fn();
  const mockSetError = jest.fn();

  const mockProps = {
    typeInput: mockTypeInput,
    user: mockUser,
    setPassword: mockSetPassword,
    setUser: mockSetUser,
    setError: mockSetError
  };

  beforeEach(() => {
    wrapper = mount(<InputElement {...mockProps}/>);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });


  it('should match the snapshot if type is first-name', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is last-name', () => {
    const wrapper = mount(
      <InputElement {...mockProps} typeInput="last-name"/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is email', () => {
    const wrapper = mount(
      <InputElement {...mockProps} typeInput="email"/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is password', () => {
    const wrapper = mount(
      <InputElement {...mockProps} typeInput="password"/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot if type is repeat-password', () => {
    const wrapper = mount(
      <InputElement {...mockProps} typeInput="repeat-password"/>
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('Change Events', () => {
    const mockEvent = {target: {value: 'Goose'}};

    it('should update local state when a change event occurs', () => {
      expect(wrapper.find('input').getDOMNode().value).toEqual('');
      wrapper.find('input').simulate('change', mockEvent);
      expect(wrapper.find('input').getDOMNode().value).toEqual('Goose');
    });

    it('should call setUser when a change event occurs', () => {
      wrapper.find('input').simulate('change', mockEvent);
      expect(mockSetUser).toHaveBeenCalled();
    });

    it('should call setPassword when a change event occurs', () => {
      const wrapper = mount(
        <InputElement {...mockProps} typeInput="repeat-password"/>
      );
      wrapper.find('input').simulate('change', mockEvent);
      expect(mockSetPassword).toHaveBeenCalled();
    });
  });
});
