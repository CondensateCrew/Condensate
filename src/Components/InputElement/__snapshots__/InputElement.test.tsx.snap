// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputElement component should match the snapshot if type is email 1`] = `
<InputElement
  setError={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setPassword={[MockFunction]}
  setUser={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  typeInput="email"
  user={
    Object {
      "email": "",
      "firstName": "",
      "lastName": "",
      "password": "",
    }
  }
>
  <label
    htmlFor="email"
  >
    email
    *
    <input
      id="email"
      onChange={[Function]}
      placeholder="email@mail.com"
      type="text"
      value=""
    />
  </label>
</InputElement>
`;

exports[`InputElement component should match the snapshot if type is first-name 1`] = `
<InputElement
  setError={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setPassword={[MockFunction]}
  setUser={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  typeInput="first-name"
  user={
    Object {
      "email": "",
      "firstName": "",
      "lastName": "",
      "password": "",
    }
  }
>
  <label
    htmlFor="first-name"
  >
    first name
    *
    <input
      id="first-name"
      onChange={[Function]}
      placeholder="Your First Name"
      type="text"
      value=""
    />
  </label>
</InputElement>
`;

exports[`InputElement component should match the snapshot if type is last-name 1`] = `
<InputElement
  setError={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setPassword={[MockFunction]}
  setUser={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  typeInput="last-name"
  user={
    Object {
      "email": "",
      "firstName": "",
      "lastName": "",
      "password": "",
    }
  }
>
  <label
    htmlFor="last-name"
  >
    last name
    *
    <input
      id="last-name"
      onChange={[Function]}
      placeholder="Your Last Name"
      type="text"
      value=""
    />
  </label>
</InputElement>
`;

exports[`InputElement component should match the snapshot if type is password 1`] = `
<InputElement
  setError={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setPassword={[MockFunction]}
  setUser={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  typeInput="password"
  user={
    Object {
      "email": "",
      "firstName": "",
      "lastName": "",
      "password": "",
    }
  }
>
  <label
    htmlFor="password"
  >
    password
    *
    <input
      id="password"
      onChange={[Function]}
      placeholder="Your Password"
      type="password"
      value=""
    />
  </label>
</InputElement>
`;

exports[`InputElement component should match the snapshot if type is repeat-password 1`] = `
<InputElement
  setError={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setPassword={
    [MockFunction] {
      "calls": Array [
        Array [
          "",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  setUser={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "email": "",
            "firstName": "",
            "lastName": "",
            "password": "",
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  typeInput="repeat-password"
  user={
    Object {
      "email": "",
      "firstName": "",
      "lastName": "",
      "password": "",
    }
  }
>
  <label
    htmlFor="repeat-password"
  >
    repeat password
    *
    <input
      id="repeat-password"
      onChange={[Function]}
      placeholder="Repeat Password"
      type="password"
      value=""
    />
  </label>
</InputElement>
`;
